### PR TITLE
ESQL: Close pages on response

### PR DESCRIPTION
**Pull request by nik9000**
This modifies ESQL to release all `Block`s when we reply to across the wire with their contents. We'd like to make all blocks `Releasable` but to do that we're really going to have to, well, release them all. This starts "at the back" of the ESQL request process.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java [hunk 1]
```diff
@@ -10,15 +10,26 @@
...
-public final class ExchangeResponse extends TransportResponse {
+public final class ExchangeResponse extends TransportResponse implements Releasable {
+    private final RefCounted counted = AbstractRefCounted.of(this::close);
     private final Page page;
     private final boolean finished;
+    /**
+     * We always use the remote exchange framwork even for local exchanges, but
+     * local exchanges shouldn't close the Page. Remote exchanges totally should
+     * as soon as they've serialized the block. This is a clever hack Nhat mentioned
+     * that can do that. But I don't like it. But it works and might unstick us.
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java:30]: I think that this is fine.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java [hunk 2]
```diff
@@ -10,15 +10,26 @@
...
+import org.elasticsearch.core.RefCounted;
+import org.elasticsearch.core.Releasable;
 import org.elasticsearch.transport.TransportResponse;
 
 import java.io.IOException;
 import java.util.Objects;
 
-public final class ExchangeResponse extends TransportResponse {
+public final class ExchangeResponse extends TransportResponse implements Releasable {
+    private final RefCounted counted = AbstractRefCounted.of(this::close);
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java:23]: Adding Releasable and RefCounted here is kinda magic, but I like it! :-)

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java:30]: :+1:

@dnhatn, if you have the time to remove this hack then please be my guest, but we might live with it for a few months.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
For those playing along at home - this doesn't actually release any breakers. Not yet. Because our pages are releasable, but we never rig them into the breakers. That is coming!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryResponse.java [hunk 3]
```diff
@@ -190,6 +192,11 @@ public String toString() {
         return Strings.toString(ChunkedToXContent.wrapAsToXContent(this));
     }
 
+    @Override
+    public void close() {
+        Releasables.close(pages.stream().map(p -> (Releasable) () -> p.releaseBlocks()).toArray(Releasable[]::new));
```
Comment (DaveCTurner) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryResponse.java]: Discussed in Slack - there's not very many `pages` normally, but this is shorter and avoids allocating an `O(pages)`-sized array too:

```suggestion
        Releasables.close(() -> Iterators.map(pages.iterator(), p -> p::releaseBlocks));
```

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java:23]: Yeah. I've copied it from other friends.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java (hunk 1 above)
Comment (dnhatn) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeResponse.java:30]: I've opened: https://github.com/elastic/elasticsearch/pull/99891

**Pull request merged**